File: src/thunks.cc
// RISC instructions are usually up to 4 bytes long, so the immediates of
// their branch instructions are naturally smaller than 32 bits.  This is
// contrary to x86-64 on which branch instructions take 4 bytes immediates
// and can jump to anywhere within PC ± 2 GiB.
//
// In fact, ARM32's branch instructions can jump only within ±16 MiB and
// ARM64's ±128 MiB, for example. If a branch target is further than that,
// we need to let it branch to a linker-synthesized code sequence that
// construct a full 32 bit address in a register and jump there. That
// linker-synthesized code is called "thunk".
//
// The function in this file creates thunks.
//
// Note that although thunks play an important role in an executable, they
// don't take up too much space in it. For example, among the clang-16's
// text segment whose size is ~300 MiB on ARM64, thunks in total occupy
// only ~30 KiB or 0.01%. Of course the number depends on an ISA; we would
// need more thunks on ARM32 whose branch range is shorter than ARM64.
// That said, the total size of thunks still isn't that much. Therefore,
// we don't need to try too hard to reduce thunk size to the absolute
// minimum.

#if MOLD_ARM32LE || MOLD_ARM32BE || MOLD_ARM64LE || MOLD_ARM64BE || \
    MOLD_PPC32 || MOLD_PPC64V1 || MOLD_PPC64V2

#include "mold.h"

#include <tbb/parallel_for.h>
#include <tbb/parallel_for_each.h>

namespace mold {

using E = MOLD_TARGET;

// We create thunks for each 25.6/3.2/6.4 MiB code block for
// ARM64/ARM32/PPC, respectively.
static constexpr i64 batch_size = branch_distance<E> / 5;

// We assume that a single thunk group is smaller than 1 MiB.
static constexpr i64 max_thunk_size = 1024 * 1024;

// We align thunks to 16 byte boundaries because many processor vendors
// recommend we align branch targets to 16 byte boundaries for performance
// reasons.
static constexpr i64 thunk_align = 16;

template <typename E>
static bool is_reachable(Context<E> &ctx, bool first_pass, InputSection<E> &isec,
                         Symbol<E> &sym, const ElfRel<E> &rel) {
  // If the target section is in the same output section but
  // hasn't got any address yet, that's unreacahble.
  InputSection<E> *isec2 = sym.get_input_section();
  if (isec2 && isec.output_section == isec2->output_section &&
      isec2->offset == -1)
    return false;

  // We don't know about the final file layout on the first pass, so
  // we assume pessimistically that all out-of-section relocations are
  // out-of-range. Excessive thunks will be removed on the second pass.
  if (first_pass) {
    if (!isec2 || isec.output_section != isec2->output_section)
      return false;

    // Even if the target is the same section, we branch to its PLT
    // if it has one. So a symbol with a PLT is also considered an
    // out-of-section reference.
    if (sym.has_plt(ctx))
      return false;
  }

  // Compute a distance between the relocated place and the symbol
  // and check if they are within reach.
  i64 S = sym.get_addr(ctx, NO_OPD);
  i64 A = get_addend(isec, rel);
  i64 P = isec.get_addr() + rel.r_offset;
  i64 val = S + A - P;
  return -branch_distance<E> <= val && val < branch_distance<E>;
}

template <typename E>
static bool needs_shim(Context<E> &ctx, Symbol<E> &sym, const ElfRel<E> &rel) {
  // Thumb and ARM B instructions cannot be converted to BX, so we
  // always have to make them jump to a thunk to switch processor mode
  // even if their destinations are within their ranges.
  if constexpr (is_arm32<E>) {
    bool is_thumb = sym.get_addr(ctx) & 1;
    return (rel.r_type == R_ARM_THM_JUMP24 && !is_thumb) ||
           (rel.r_type == R_ARM_JUMP24 && is_thumb) ||
           (rel.r_type == R_ARM_PLT32 && is_thumb);
  }

  // On PowerPC, all PLT calls go through range extension thunks.
  if constexpr (is_ppc32<E> || is_ppc64v1<E>)
    return sym.has_plt(ctx);

  // PowerPC before Power9 lacks PC-relative load/store instructions.
  // Functions compiled for Power9 or earlier assume that r2 points to
  // GOT+0x8000, while those for Power10 uses r2 as a scratch register.
  // We need to a thunk to recompute r2 for interworking.
  if constexpr (is_ppc64v2<E>)
    return sym.has_plt(ctx) ||
           (rel.r_type == R_PPC64_REL24 && !sym.esym().ppc64_preserves_r2()) ||
           (rel.r_type == R_PPC64_REL24_NOTOC && sym.esym().ppc64_uses_toc());
  return false;
}

template <>
void OutputSection<E>::create_range_extension_thunks(Context<E> &ctx,
                                                     bool first_pass) {
  std::span<InputSection<E> *> m = members;
  if (m.empty())
    return;

  // Initialize input sections with a dummy offset so that we can
  // distinguish sections that have got an address with the one who
  // haven't.
  for (InputSection<E> *isec : m)
    isec->offset = -1;
  thunks.clear();

  // We create thunks from the beginning of the section to the end.
  // We manage progress using four offsets which increase monotonically.
  // The locations they point to are always A <= B <= C <= D.
  //
  // Input sections between B and C are in the current batch.
  //
  // A is the input section with the smallest address than can reach
  // from the current batch.
  //
  // D is the input section with the largest address such that the thunk
  // is reachable from the current batch if it's inserted at D.
  //
  //  ................................ <input sections> ............
  //     A    B    C    D
  //                    ^ We insert a thunk for the current batch just before D
  //          <--->       The current batch, which is smaller than BATCH_SIZE
  //     <-------->       Smaller than BRANCH_DISTANCE
  //          <-------->  Smaller than BRANCH_DISTANCE
  //     <------------->  Reachable from the current batch
  i64 a = 0;
  i64 b = 0;
  i64 c = 0;
  i64 d = 0;
  i64 offset = 0;

  // The smallest thunk index that is reachable from the current batch.
  i64 t = 0;

  while (b < m.size()) {
    // Move D foward as far as we can jump from B to a thunk at D.
    auto d_thunk_end = [&] {
      u64 d_end = align_to(offset, 1 << m[d]->p2align) + m[d]->sh_size;
      return align_to(d_end, thunk_align) + max_thunk_size;
    };

    while (d < m.size() &&
           (b == d || d_thunk_end() <= m[b]->offset + branch_distance<E>)) {
      offset = align_to(offset, 1 << m[d]->p2align);
      m[d]->offset = offset;
      offset += m[d]->sh_size;
      d++;
    }

    // Move C forward so that C is apart from B by BATCH_SIZE. We want
    // to make sure that there's at least one section between B and C
    // to ensure progress.
    c = b + 1;
    while (c < d && m[c]->offset + m[c]->sh_size < m[b]->offset + batch_size)
      c++;

    // Move A forward so that A is reachable from C.
    i64 c_offset = (c == d) ? offset : m[c]->offset;
    while (a < b && m[a]->offset + branch_distance<E> < c_offset)
      a++;

    // Erase references to out-of-range thunks.
    for (; t < thunks.size() && thunks[t]->offset < m[a]->offset; t++)
      for (Symbol<E> *sym : thunks[t]->symbols)
        sym->flags = 0;

    // Create a new thunk and place it at D.
    offset = align_to(offset, thunk_align);
    thunks.emplace_back(std::make_unique<Thunk<E>>(*this, offset));
    Thunk<E> &thunk = *thunks.back();

    // Scan relocations between B and C to collect symbols that need
    // entries in the new thunk.
    std::mutex mu;

    tbb::parallel_for(b, c, [&](i64 i) {
      InputSection<E> &isec = *m[i];

      for (const ElfRel<E> &rel : isec.get_rels(ctx)) {
        if (!is_func_call_rel(rel))
          continue;

        // Skip if the symbol is undefined. apply_reloc() will report an error.
        Symbol<E> &sym = *isec.file.symbols[rel.r_sym];
        if (!sym.file)
          continue;

        // Skip if we can directly branch to the destination.
        if (is_reachable(ctx, first_pass, isec, sym, rel) &&
            !needs_shim(ctx, sym, rel))
          continue;

        // Add the symbol to the current thunk if it's not added already.
        if (!sym.flags.test_and_set()) {
          std::scoped_lock lock(mu);
          thunk.symbols.push_back(&sym);
        }
      }
    });

    // Sort symbols added to the thunk to make the output deterministic.
    ranges::sort(thunk.symbols, {}, [](Symbol<E> *x) {
      return std::tuple{x->file->priority, x->sym_idx};
    });

    // Now that we know the number of symbols in the thunk, we can compute
    // the thunk's size.
    assert(thunk.size() < max_thunk_size);
    offset += thunk.size();

    // Move B forward to point to the begining of the next batch.
    b = c;
  }

  for (; t < thunks.size(); t++)
    for (Symbol<E> *sym : thunks[t]->symbols)
      sym->flags = 0;

  this->shdr.sh_size = offset;
}

template <>
void remove_redundant_thunks(Context<E> &ctx) {
  Timer t(ctx, "remove_redundant_thunks");
  set_osec_offsets(ctx);

  for (Chunk<E> *chunk : ctx.chunks)
    if (OutputSection<E> *osec = chunk->to_osec())
      if (osec->shdr.sh_flags & SHF_EXECINSTR)
        osec->create_range_extension_thunks(ctx, false);
}

// When applying relocations, we want to know the address in a reachable
// range extension thunk for a given symbol. Doing it by scanning all
// reachable range extension thunks is too expensive.
//
// In this function, we create a list of all addresses in range extension
// thunks for each symbol, so that it is easy to find one.
//
// Note that thunk_addrs must be sorted for binary search.
template <>
void gather_thunk_addresses(Context<E> &ctx) {
  Timer t(ctx, "gather_thunk_addresses");

  std::vector<OutputSection<E> *> sections;
  for (Chunk<E> *chunk : ctx.chunks)
    if (OutputSection<E> *osec = chunk->to_osec())
      sections.push_back(osec);

  ranges::stable_sort(sections, {}, [](OutputSection<E> *x) {
    return x->shdr.sh_addr;
  });

  for (OutputSection<E> *osec : sections) {
    for (std::unique_ptr<Thunk<E>> &thunk : osec->thunks) {
      for (i64 i = 0; i < thunk->symbols.size(); i++) {
        Symbol<E> &sym = *thunk->symbols[i];
        sym.add_aux(ctx);
        ctx.symbol_aux[sym.aux_idx].thunk_addrs.push_back(thunk->get_addr(i));
      }
    }
  }
}

} // namespace mold

#endif

File: src/thunks.rs
//! RISC instructions are usually up to 4 bytes long, so the immediates of
//! their branch instructions are naturally smaller than 32 bits.  This is
//! contrary to x86-64 on which branch instructions take 4 bytes immediates
//! and can jump to anywhere within PC ± 2 GiB.
//!
//! In fact, ARM32's branch instructions can jump only within ±16 MiB and
//! ARM64's ±128 MiB, for example. If a branch target is further than that,
//! we need to let it branch to a linker-synthesized code sequence that
//! construct a full 32 bit address in a register and jump there. That
//! linker-synthesized code is called "thunk".
//!
//! The function in this file creates thunks.
//!
//! Thunk size varies widely across programs. In an ARM64 build of Clang 16,
//! thunks occupy about 30 KiB (0.01%) of a ~300 MiB text section, compared
//! with about 12.5 MiB (2.5%) of a ~500 MiB text section in TensorFlow.
//!
//! Range extension thunks.
//!
//! RISC branch instructions have small immediates: ARM32 branches reach
//! ±16 MiB and ARM64 ones ±128 MiB, unlike x86-64's 32-bit displacements.
//! A call whose target is further away is redirected to a thunk, a
//! linker-synthesized code sequence between input sections that loads the
//! full address into a register and jumps there.
//!
//! Thunks are created in two passes. Before addresses are known, every
//! out-of-section call is pessimistically assumed to need a thunk; once
//! the layout is fixed, the entries that turned out to be unneeded are
//! removed. Sections only shrink in the second pass, so no existing
//! reference to a thunk goes out of range because of it.

use std::cell::UnsafeCell;

use rayon::prelude::*;

use crate::arch::{Arch, Family};
use crate::context::Context;
use crate::elf::*;
use crate::input_sections::InputSection;
use crate::output_chunks::output_section::OutputSection;
use crate::output_chunks::{ChunkId, OutputSectionId};
use crate::symbol::{AddrFlags, Symbol, SymbolId};
use crate::util::align_to;

/// A block of branch stubs placed between input sections.
#[derive(Debug)]
pub struct Thunk {
    /// Offset within the output section.
    pub offset: u64,
    pub symbols: Vec<SymbolId>,
    /// Offset of each stub within the thunk; the last entry is the size.
    pub offsets: Vec<u64>,
    pub name: String,
}

impl Thunk {
    pub fn size(&self) -> u64 {
        self.offsets.last().copied().unwrap_or(0)
    }

    pub fn addr(&self, osec: &OutputSection) -> u64 {
        osec.hdr.shdr.sh_addr + self.offset
    }

    /// The entry offsets of a thunk with fixed-size entries.
    pub fn fixed_offsets<E: Arch>(&self) -> Vec<u64> {
        let layout = E::THUNK.expect("target without thunks");
        (0..=self.symbols.len())
            .map(|i| layout.header_size + i as u64 * layout.entry_size)
            .collect()
    }
}

/// A section offset that hasn't been assigned yet.
const UNPLACED: u64 = u64::MAX;

/// We create thunks for each 32/8/16 MiB code block for
/// ARM64/ARM32/PPC, respectively.
fn batch_size<E: Arch>() -> u64 {
    let mib = match E::FAMILY {
        Family::Arm64 => 32,
        Family::Arm32 => 8,
        _ => 16,
    };
    mib << 20
}

/// We assume that a single thunk group is smaller than 16/4/8 MiB
/// for ARM64/ARM32/PPC, respectively.
fn max_thunk_size<E: Arch>() -> u64 {
    batch_size::<E>() / 2
}

/// Power10 prefixed instructions must not cross a 64-byte boundary.
/// Aligning each thunk group to a 8-byte boundary guarantees that.
const THUNK_ALIGN: u64 = 8;

/// The thunk symbols collected by each Rayon worker. This is the equivalent
/// of C++ mold's `enumerable_thread_specific<std::vector<Symbol *>>`.
struct ThunkSymbolBins(Vec<UnsafeCell<Vec<SymbolId>>>);

// SAFETY: a Rayon worker has a unique index and executes at most one closure
// at a time. The only caller does not invoke nested parallel work while using
// its bin; the final merge happens after the parallel traversal has joined.
unsafe impl Sync for ThunkSymbolBins {}

impl ThunkSymbolBins {
    fn new() -> ThunkSymbolBins {
        ThunkSymbolBins(
            (0..=rayon::current_num_threads())
                .map(|_| UnsafeCell::new(Vec::new()))
                .collect(),
        )
    }

    #[inline]
    fn push(&self, sym: SymbolId) {
        let fallback = self.0.len() - 1;
        let i = rayon::current_thread_index()
            .unwrap_or(fallback)
            .min(fallback);
        // SAFETY: the Sync invariant above gives this worker exclusive access
        // to its indexed vector for the duration of this non-nested closure.
        unsafe { &mut *self.0[i].get() }.push(sym);
    }

    fn into_vec(self) -> Vec<SymbolId> {
        let mut bins: Vec<Vec<SymbolId>> = self.0.into_iter().map(UnsafeCell::into_inner).collect();
        let mut symbols = Vec::with_capacity(bins.iter().map(Vec::len).sum());
        for bin in &mut bins {
            symbols.append(bin);
        }
        symbols
    }
}

/// Whether a call needs a thunk. On the first pass, before addresses are
/// known, every call out of the section is assumed to need one.
#[inline(always)]
fn requires_thunk<E: Arch>(
    ctx: &Context<E>,
    isec: &InputSection,
    rel: &ElfRel,
    sym: &Symbol,
    first_pass: bool,
) -> bool {
    if !rel.is_func_call::<E>() {
        return false;
    }

    if first_pass {
        // On the first pass, we pessimistically assume that all out-of-section
        // relocations are out of range.
        match sym.input_section_ref() {
            Some(target) if target.output_section == isec.output_section => {
                // If the target section is in the same output section but
                // hasn't got any address yet, that's unreacahble.
                if target.offset() == UNPLACED {
                    return true;
                }
            }
            _ => return true,
        }

        // Even if the target is the same section, we branch to its PLT
        // if it has one. So a symbol with a PLT is also considered an
        // out-of-section reference.
        if sym.has_plt(&ctx.symbols) {
            return true;
        }
    }

    if E::always_needs_thunk(ctx, sym, rel) {
        return true;
    }

    // Compute a distance between the relocated place and the symbol
    // and check if they are within reach.
    let s = sym.addr_with(ctx, AddrFlags::NO_OPD) as i64;
    let a = isec.rel_addend::<E>(rel);
    let p = (isec.addr(ctx) + rel.r_offset) as i64;
    let val = s.wrapping_add(a).wrapping_sub(p);
    val < -E::branch_distance() || E::branch_distance() <= val
}

/// The executable output sections, in output order.
fn executable_sections<E: Arch>(ctx: &Context<E>) -> Vec<OutputSectionId> {
    ctx.chunks
        .iter()
        .filter_map(|&id| match id {
            ChunkId::Output(osec)
                if ctx.output_sections[osec.index()].hdr.shdr.sh_flags & SHF_EXECINSTR as u64
                    != 0 =>
            {
                Some(osec)
            }
            _ => None,
        })
        .collect()
}

/// We create thunks from the beginning of the section to the end.
/// We manage progress using four offsets which increase monotonically.
/// The locations they point to are always A <= B <= C <= D.
///
/// Input sections between B and C are the current batch.
///
/// A is the input section with the smallest address than can reach
/// from the current batch.
///
/// D is the input section with the largest address such that the thunk
/// is reachable from the current batch if it's inserted at D.
///
///  ................................ <input sections> ............
///     A    B    C    D
///                    ^ We insert a thunk for the current batch just before D
///          <--->       The current batch, which is smaller than BATCH_SIZE
///     <-------->       Smaller than BRANCH_DISTANCE
///          <-------->  Smaller than BRANCH_DISTANCE
///     <------------->  Reachable from the current batch
pub fn create_range_extension_thunks<E: Arch>(ctx: &mut Context<E>, id: OutputSectionId) {
    let members = std::mem::take(&mut ctx.output_sections[id.index()].members);
    if members.is_empty() {
        return;
    }

    // Initialize input sections with a dummy offset so that we can
    // distinguish sections whose addresses have been assigned from those
    // whose addresses have not.
    members
        .par_iter()
        .for_each(|&member| ctx.input_section(member).set_offset(UNPLACED));

    let distance = E::branch_distance() as u64;
    let batch = batch_size::<E>();
    let max_thunk = max_thunk_size::<E>();
    let n = members.len();

    let mut thunks: Vec<Thunk> = Vec::new();
    let (mut a, mut b, mut d) = (0usize, 0usize, 0usize);
    let mut offset = 0u64;
    // The smallest thunk index that is reachable from the current batch.
    let mut t = 0usize;

    while b < n {
        // Move D foward as far as we can jump from B to a thunk at D.
        while d < n {
            let sec = ctx.input_section(members[d]);
            let (p2align, size) = (sec.p2align(), sec.sh_size);
            if b != d {
                let thunk_end =
                    align_to(align_to(offset, 1 << p2align) + size, THUNK_ALIGN) + max_thunk;
                if thunk_end > ctx.input_section(members[b]).offset() + distance {
                    break;
                }
            }
            offset = align_to(offset, 1 << p2align);
            sec.set_offset(offset);
            offset += size;
            d += 1;
        }

        // Find the end of the current batch. Section end addresses are sorted,
        // so use binary search. Starting from B + 1 guarantees progress.
        let b_offset = ctx.input_section(members[b]).offset();
        let c = b
            + 1
            + members[b + 1..d].partition_point(|&member| {
                let sec = ctx.input_section(member);
                sec.offset() + sec.sh_size < b_offset + batch
            });

        // Find the first section that is within branch range of C.
        let c_offset = if c == d {
            offset
        } else {
            ctx.input_section(members[c]).offset()
        };
        a += members[a..b].partition_point(|&member| {
            ctx.input_section(member).offset() < c_offset.saturating_sub(distance)
        });

        // Erase references to out-of-range thunks.
        while t < thunks.len() && thunks[t].offset < ctx.input_section(members[a]).offset() {
            for &sym in &thunks[t].symbols {
                ctx.symbols[sym].unmark();
            }
            t += 1;
        }

        // Create a new thunk and place it at D.
        offset = align_to(offset, THUNK_ALIGN);
        let symbol_bins = ThunkSymbolBins::new();
        {
            let ctx: &Context<E> = ctx;
            // Scan relocations between B and C to collect symbols that need
            // entries in the new thunk.
            members[b..c].par_iter().for_each(|&member| {
                let isec = ctx.input_section(member);
                let file = &ctx.objs[isec.file.index()];
                for rel in isec.rels::<E>(file) {
                    if !rel.is_func_call::<E>() {
                        continue;
                    }
                    let id = file.base.symbols[rel.r_sym as usize];
                    let sym = &ctx.symbols[id];
                    if requires_thunk(ctx, isec, &rel, sym, true) && sym.mark() {
                        symbol_bins.push(id);
                    }
                }
            });
        }
        // Add symbols to the thunk
        let mut thunk = Thunk {
            offset,
            symbols: symbol_bins.into_vec(),
            offsets: Vec::new(),
            name: String::new(),
        };
        // Now that we know the number of symbols in the thunk, we can compute
        // the thunk's size.
        thunk.offsets = thunk.fixed_offsets::<E>();
        debug_assert!(thunk.size() < max_thunk);
        offset += thunk.size();
        thunks.push(thunk);

        // Move B forward to point to the begining of the next batch.
        b = c;
    }

    // Clear marks for thunks that are still reachable from the last batch.
    for thunk in &thunks[t..] {
        for &sym in &thunk.symbols {
            ctx.symbols[sym].unmark();
        }
    }

    // Sort symbols for deterministic output.
    {
        let ctx: &Context<E> = ctx;
        thunks.par_iter_mut().for_each(|thunk| {
            thunk.symbols.sort_by_key(|&id| {
                let sym = &ctx.symbols[id];
                (sym.file().map_or(0, |f| ctx.file(f).priority), sym.sym_idx)
            });
        });
    }

    let osec = &mut ctx.output_sections[id.index()];
    osec.hdr.shdr.sh_size = offset;
    osec.thunks = thunks;
    osec.members = members;
}

/// create_range_extension_thunks() creates thunks with a pessimistic
/// assumption that all out-of-section references are out of range.
/// After computing output section addresses, we revisit all thunks to
/// remove unneeded entries from them.
///
/// We create more thunks than necessary and then eliminate some of
/// them later, instead of just creating thunks at this stage. This is
/// because we can safely shrink sections after assigning addresses to
/// them without worrying about making existing references to thunks go
/// out of range. On the other hand, if we insert thunks after
/// assigning addresses to sections, references to thunks could become
/// out of range due to the new extra gaps for thunks. Thus, the
/// creation of thunks is a two-pass process.
pub fn remove_redundant_thunks<E: Arch>(ctx: &mut Context<E>) {
    let _t = ctx.timer("remove_redundant_thunks");
    // Gather output executable sections
    let sections = executable_sections(ctx);

    // Mark all symbols that actually need range extension thunks
    {
        let ctx: &Context<E> = ctx;
        for &id in &sections {
            ctx.output_sections[id.index()]
                .members
                .par_iter()
                .for_each(|&m| {
                    let isec = ctx.input_section(m);
                    let file = &ctx.objs[isec.file.index()];
                    for rel in isec.rels::<E>(file) {
                        if !rel.is_func_call::<E>() {
                            continue;
                        }
                        let sym = &ctx.symbols[file.base.symbols[rel.r_sym as usize]];
                        if !sym.is_marked() && requires_thunk(ctx, isec, &rel, sym, false) {
                            sym.mark();
                        }
                    }
                });
        }
    }

    for &id in &sections {
        let mut thunks = std::mem::take(&mut ctx.output_sections[id.index()].thunks);
        {
            let ctx: &Context<E> = ctx;
            let osec = &ctx.output_sections[id.index()];
            // Remove symbols from thunks if they don't actually need range
            // extension thunks
            thunks.par_iter_mut().for_each(|thunk| {
                thunk.symbols.retain(|&sym| ctx.symbols[sym].is_marked());
                thunk.offsets = E::thunk_offsets(ctx, thunk, thunk.addr(osec));
            });
        }

        // Recompute section sizes
        let members = ctx.output_sections[id.index()].members.clone();
        let (mut mi, mut ti) = (0, 0);
        let mut offset = 0;
        while mi < members.len() || ti < thunks.len() {
            let member_first = mi < members.len()
                && (ti >= thunks.len()
                    || ctx.input_section(members[mi]).offset() < thunks[ti].offset);
            if member_first {
                let sec = ctx.input_section(members[mi]);
                offset = align_to(offset, 1 << sec.p2align());
                sec.set_offset(offset);
                offset += sec.sh_size;
                mi += 1;
            } else {
                offset = align_to(offset, THUNK_ALIGN);
                thunks[ti].offset = offset;
                offset += thunks[ti].size();
                ti += 1;
            }
        }
        let osec = &mut ctx.output_sections[id.index()];
        osec.hdr.shdr.sh_size = offset;
        osec.thunks = thunks;
    }
}

/// When applying relocations, we want to know the address in a reachable
/// range extension thunk for a given symbol. Doing it by scanning all
/// reachable range extension thunks is too expensive.
///
/// In this function, we create a list of all addresses in range extension
/// thunks for each symbol, so that it is easy to find one.
///
/// Note that thunk_addrs must be sorted for binary search.
pub fn gather_thunk_addresses<E: Arch>(ctx: &mut Context<E>) {
    let _t = ctx.timer("gather_thunk_addresses");
    let mut sections = executable_sections(ctx);
    sections.sort_by_key(|id| ctx.output_sections[id.index()].hdr.shdr.sh_addr);

    for id in sections {
        let osec = &ctx.output_sections[id.index()];
        let entries: Vec<(SymbolId, u64)> = osec
            .thunks
            .iter()
            .flat_map(|thunk| {
                let base = thunk.addr(osec);
                thunk
                    .symbols
                    .iter()
                    .enumerate()
                    .map(move |(i, &sym)| (sym, base + thunk.offsets[i]))
            })
            .collect();
        for (sym, addr) in entries {
            ctx.symbols.add_thunk_addr(sym, addr);
        }
    }
}

/// Writes a thunk's stubs into the output section buffer.
pub fn copy_buf<E: Arch>(ctx: &Context<E>, osec: &OutputSection, thunk: &Thunk, buf: &mut [u8]) {
    debug_assert_eq!(buf.len(), thunk.size() as usize);
    E::write_thunk(ctx, thunk, thunk.addr(osec), buf);
}
